documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,243 @@
+		// in relabeling rules don't have to consider tag positions.
+		var tags = "," + strings.Join(node.ServiceTags, ",") + ","
+
+		// If the service address is not empty it should be used instead of the node address
+		// since the service may be registered remotely through a different node.
+		var addr string
+		if node.ServiceAddress != "" {
+			addr = net.JoinHostPort(node.ServiceAddress, fmt.Sprintf("%d", node.ServicePort))
+		} else {
+			addr = net.JoinHostPort(node.Address, fmt.Sprintf("%d", node.ServicePort))
+		}
+
+		target := model.LabelSet{model.AddressLabel: model.LabelValue(addr)}
+		labels := model.LabelSet{
+			model.AddressLabel:                   model.LabelValue(addr),
+			model.LabelName(addressLabel):        model.LabelValue(node.Address),
+			model.LabelName(nodeLabel):           model.LabelValue(node.Node),
+			model.LabelName(tagsLabel):           model.LabelValue(tags),
+			model.LabelName(serviceAddressLabel): model.LabelValue(node.ServiceAddress),
+			model.LabelName(servicePortLabel):    model.LabelValue(strconv.Itoa(node.ServicePort)),
+			model.LabelName(serviceIDLabel):      model.LabelValue(node.ServiceID),
+		}
+		tgroup.Labels = labels
+
+		// Add all key/value pairs from the node's metadata as their own labels.
+		for k, v := range node.NodeMeta {
+			name := strutil.SanitizeLabelName(k)
+			tgroup.Labels[model.LabelName(name)] = model.LabelValue(v)
+		}
+
+		tgroup.Targets = append(tgroup.Targets, target)
+	}
+	return &tgroup, nil
+}
+
+// Note: you must implement this function for your discovery implementation as part of the
+// Discoverer interface. Here you should query your SD for it's list of known targets, determine
+// which of those targets you care about (for example, which of Consuls known services do you want
+// to scrape for metrics), and then send those targets as a target.TargetGroup to the ch channel.
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	for c := time.Tick(time.Duration(d.refreshInterval) * time.Second); ; {
+		var srvs map[string][]string
+		resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/services", d.address))
+
+		if err != nil {
+			level.Error(d.logger).Log("msg", "Error getting services list", "err", err)
+			time.Sleep(time.Duration(d.refreshInterval) * time.Second)
+			continue
+		}
+
+		dec := json.NewDecoder(resp.Body)
+		err = dec.Decode(&srvs)
+		resp.Body.Close()
+		if err != nil {
+			level.Error(d.logger).Log("msg", "Error reading services list", "err", err)
+			time.Sleep(time.Duration(d.refreshInterval) * time.Second)
+			continue
+		}
+
+		var tgs []*targetgroup.Group
+		for name := range srvs {
+			if name == "consul" {
+				continue
+			}
+			resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/service/%s", d.address, name))
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
+				continue
```

Comment:
This should abort this cycle, not continue on to the next service.

Reply:
how come? we want to assume all `service/<serviceName>` endpoints will be down if one fails?

I was thinking that between the time we query for all services and the time we query for each services nodes that a service may be removed, so querying it's endpoint would fail but the rest may succeed

Reply:
It'd better to have stale data than incomplete data.

Reply:
In this case would it really be incomplete? A service is no longer available

Reply:
How would we distinguish a service no longer existing from a timeout?